Avoid scheduling rendering updates for SVG animations on elements

nullhook opened the pull request:
#### b70eabc441428cc812dd279f5f2df22b237bc523
<pre>
Avoid scheduling rendering updates for SVG animations on elements
without renderers
<a href="https://bugs.webkit.org/show_bug.cgi?id=298217">https://bugs.webkit.org/show_bug.cgi?id=298217</a>
<a href="https://rdar.apple.com/159647563">rdar://159647563</a>

Reviewed by Antoine Quint.

When SVG elements with SMIL animations (like animateTransform) are
inside containers with hidden attribute, the legacy SVG engine was
scheduling full page rendering updates every animation frame despite
the elements having no renderer.

Now when animations run on elements without renderers, we skip
scheduling rendering updates entirely.

* LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates-expected.txt: Added.
* LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html: Added.
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::svgAttributeChanged):

Canonical link: <a href="https://commits.webkit.org/304744@main">https://commits.webkit.org/304744@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bcc831e50506ed50cb2d4479bfc46940c5a279ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136394 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47674 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144106 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a557c01c-7e6b-4dc6-9639-4b2943aa8f41) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138266 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104294 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/00232f2d-0ff4-4f6e-bca4-1178333a7e3f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6870 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122208 "Found 1 new API test failure: TestWebKitAPI.InWindowFullscreen.IFrameDocument (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85130 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b38fd90c-0fdb-4fb5-9ee2-ebd448b65a85) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6517 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4178 "Found 1 new API test failure: TestWebKitAPI.FormValidation.PresentingFormValidationUIWithoutViewControllerDoesNotCrash (failure)") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4697 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115811 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146850 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8433 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40982 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112630 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8450 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7080 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112977 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28689 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6453 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118517 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62420 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8481 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36569 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8199 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72040 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8421 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8273 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html:
```diff
@@ -0,0 +1,74 @@
+<html>
+
+<head>
+    <script src="../../resources/js-test-pre.js"></script>
```
For new tests - we prefer to use `js-test.js`. So you can remove `pre` and `post`.

graouts commented on LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html:
```diff
@@ -0,0 +1,74 @@
+            if (updateCount > 5)
+                debug(`Saw ${updateCount} rendering updates (expected <= 5)`);
+            else
+                debug(`Saw ${updateCount} rendering updates`);
```
This logging leads to flakiness per EWS. You should get rid of this, or only keep the failing case.

graouts commented on LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html:
```diff
@@ -0,0 +1,74 @@
+            let renderingUpdateCount = 0;
+            do {
+                renderingUpdateCount = internals.renderingUpdateCount();
+                await sleepFor(32);
```
I expect 32ms is chosen to match two frames? But 33ms is a closer value… and perhaps 34 would make more sense if you round up. At any rate, it would be good to explain the rationale here.

nullhook replied:
good point. totally forgot the rounding of `1 frame = ~16.67ms` on a 60fps display

graouts commented on LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html:
```diff
@@ -0,0 +1,74 @@
+            await waitForRenderingUpdateCountToSettle();
+
+            // wait for 2s while the animation runs (2 full cycles at 1s each)
+            await sleepFor(2000);
```
That's a pretty long time for a test… is it necessary to run this long? Surely the equivalent of 10 rendering frames or so should be enough to determine if we're scheduling rendering updates here.

nullhook replied:
revised it to 250ms

graouts commented on LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html:
```diff
@@ -0,0 +1,69 @@
+
+            internals.setSpeculativeTilingDelayDisabledForTesting(true);
+            
+            await waitForRenderingUpdateCountToSettle();
```
With the call to `setSpeculativeTilingDelayDisabledForTesting` just above, is that part necessary?

nullhook replied:
something to do with determinism when reading rendering updates. i think it adds more rendering updates and result in flakiness? not 100% sure.

nullhook committed:
Avoid scheduling rendering updates for SVG animations on elements
without renderers
https://bugs.webkit.org/show_bug.cgi?id=298217
rdar://159647563

Reviewed by Antoine Quint.

When SVG elements with SMIL animations (like animateTransform) are
inside containers with hidden attribute, the legacy SVG engine was
scheduling full page rendering updates every animation frame despite
the elements having no renderer.

Now when animations run on elements without renderers, we skip
scheduling rendering updates entirely.

* LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates-expected.txt: Added.
* LayoutTests/svg/animations/animateTransform-hidden-no-rendering-updates.html: Added.
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::svgAttributeChanged):

Canonical link: https://commits.webkit.org/304744@main
